R studio demo course

Rymik19 opened the pull request:
This demo course is based on Duke University's Statistical Science Introductory Data Science courses: [DukeStatSci/introds](https://github.com/DukeStatSci/introds). 

It includes general instructions (*README.md*) to course creation and how this particular course was created. It contains a `build-and-run-course.py` which can be simply executed to build and test the R Studio image with imported course content. Additionally, it uses an initialization script and adds persistent files  (*/renv* and */slides*) to the `/work` directory that are accessible in different classes.

JHavstein committed:
Create create-new-course.py

FedericaLoVerso committed:
add templates

JHavstein committed:
Add required format for -r

JHavstein committed:
Merge branch 'add-course-setup' of https://github.com/SDU-eScience/UCloud-Courses into add-course-setup

JHavstein committed:
Add course start date format check

The user has to give the start date of the course (the -r option) in the format: YYYY-MM-DD.

JHavstein committed:
Small change

JHavstein committed:
Add elaborate error checking for -release

JHavstein committed:
Small change

JHavstein committed:
Fix small error

JHavstein committed:
Add code annotation

JHavstein committed:
Fix spacing

JHavstein committed:
Add a little string formatting

Cosmetic changes only.

JHavstein committed:
Fix error in outline of folder structure

JHavstein committed:
Prepare for editing of template files

The contents of the template files (in /templates) have to be edited based on information from the user such as course name, base image etc. This is done with a simple read-edit-write procedure.

TODO: Implement the 'edit' step.

JHavstein committed:
Add script documentation

JHavstein committed:
Add error checking and more format enforcement

Added error checking to ensure that the user is in the correct working directory. This is a requirement when creating the course folder structure.

Added an enforcement on the course name. Specifically, if the user enters a course name that contains whitespaces, the course name is stripped of excess whitespace, and remaining whitespace is replaced with dashes.

JHavstein committed:
Remove print statement

JHavstein committed:
Add to error messages

JHavstein committed:
Add cleanup in case of error

If an error happens occurs after the course's folder structure has been created, the course folder is deleted recursively starting from course_root_dir.

JHavstein committed:
Minor addition

JHavstein committed:
Add more format enforcement

Remove all leading and trailing whitespaces from args.name. This will ensure that a course does not start or end with a dash due to whitespace substitution.

JHavstein committed:
Add base image checker

Added functionality to map from args.baseimage to the newest version of the actual base image that will be used for any given course.

Also added a prompt asking the user if the program should delete all the course folders in case the program exits on an error. This is to minimize the risk of accidental folder deletion.

JHavstein committed:
Finalize get_baseimage_name()

Also delete course that was accidentally pushed in previous commit.

JHavstein committed:
Remove tag suffix for VDEs

The '-xfce' suffix is hardcoded into the Dockerfiles instead.

JHavstein committed:
Edit template files before writing to course folder

JHavstein committed:
Remove cleanup on Error run

With this cleanup, we risked that teachers accidentally deleted other teacher's courses if they made a course with the same name as an existing course.

JHavstein committed:
Add error check on course name availability

JHavstein committed:
Add text substitution in README template

JHavstein committed:
Remove ARG TAG from Dockerfile templates

The tag is substituted into the Dockerfile templates from create-new-course.py, so ARG TAG is not used. Furthermore, the TAG is substituted with the actual tag, so `ARG TAG` will not be present in the templates that are copied into the course folders because `TAG` will be replaced with the actual tag.

JHavstein committed:
Create requirements.txt

The remaining modules that are needed to execute create-new-course.py is parts of the Python standard library.

JHavstein committed:
Add Python version as comment

Python 3.10.14.

JHavstein committed:
Remove __version__

Because not really meaningful here.

JHavstein committed:
Remove __author__

Not really meaningful here.

JHavstein committed:
Minor additions

FedericaLoVerso committed:
change templ

FedericaLoVerso committed:
change script n

JHavstein committed:
Change folders

- Moved README.md to the course root directory.
- Created /test and /logo is the course root directory
- Change 'startapp' to 'startcourse'

JHavstein committed:
Remove __status__

Not really meaningful.

JHavstein committed:
Enforce that <course name> must be lowercase

JHavstein committed:
Remove unused function

FedericaLoVerso committed:
update templates

JHavstein committed:
Add directory

JHavstein committed:
Incorporate suggestions from code review

Commit:
Run create-new-course-script

Commit:
clone from repo https://github.com/rbroc/NLP-AU-23

Commit:
test course module 1

Commit:
move requirements.txt

Commit:
template for makefile

Commit:
let Makefile take a -c [class] argument

Commit:
remove Makefile

JHavstein committed:
Add build-docker-image.py

Script for building docker images from course Dockerfiles. Should be portable across operating systems but currently only tested on macOS. Also update requirements.txt.

JHavstein committed:
Fix typo

JHavstein committed:
Add info about Docker daemon

Specifically, the script can only run if a Docker daemon is running.

JHavstein committed:
Add maintainer info

JHavstein committed:
Update .gitignore

Ignore .DS_Store everywhere in repo.

JHavstein committed:
Add uni and course code info

Courses/ now contains subfolders for each university (to being with only AAU, AU, and SDU). Courses for a given university are then placed in the given university folder. Also added the course code (from university course description) to the course folder name (exact naming convention to be decided). University and course code are added as command line argument (-u and -c, respectively).

JHavstein committed:
Add subfolder for each university

Only AAU, AU, and SDU to begin with.

JHavstein committed:
Update build-docker-image.py

Add args -u and -c to be able to match course as created using create-new-course.py.
Also changed the image tag.

JHavstein committed:
Update explanatory comment

Add explanation of -c and -u args.

Commit:
WIP. Download course materials from external repo

Commit:
Cleanup and error handling

Commit:
dependencies handled

Commit:
save course resources from external repo in '/work'

Commit:
delete unnecessary print

Commit:
newline at ends of files

Commit:
Merge remote-tracking branch 'origin/add-course-setup' into nlpDemoCourse

Commit:
Follow new structure after rebase: move demo course into AU folder

Commit:
Follow new structure after rebase: move demo course into AU folder

JHavstein committed:
Add remaining universities + other

JHavstein committed:
Ensure uniformity in naming

Rymik19 committed:
Adds new rstudio demo course

Rymik19 committed:
au folder index change

Rymik19 committed:
Fix docker build image

Commit:
Merge remote-tracking branch 'origin/add-course-setup' into nlpDemoCourse

Commit:
Merge remote-tracking branch 'origin/add-course-setup' into nlpDemoCourse

JHavstein committed:
Update scripts/build-docker-image.py

Co-authored-by: Jes Elgin Jensen <146733701+jeselginAU@users.noreply.github.com>

JHavstein committed:
Update scripts/create-new-course.py

Co-authored-by: Jes Elgin Jensen <146733701+jeselginAU@users.noreply.github.com>

JHavstein committed:
Remove comment

Rymik19 committed:
Fixes build docker image script -u choices and removes nlp course from branch

Commit:
Merge branch 'nlpDemoCourse' of github.com:SDU-eScience/UCloud-Courses into nlpDemoCourse

Commit:
Merge branch 'nlpDemoCourse' into rStudioDemoCourse

Commit:
Merge remote-tracking branch 'origin/add-course-setup' into nlpDemoCourse

Commit:
Merge remote-tracking branch 'origin/nlpDemoCourse' into rStudioDemoCourse

Commit:
update README

Rymik19 committed:
added script instructions to README.

Rymik19 committed:
remove unnecessary files

Rymik19 committed:
adds requirements script for launching rstudio, updates note on commands for py scripts

Rymik19 committed:
adds force download flag and check if files already exist to skip download

Rymik19 committed:
prints downloaded file name

Commit:
explain parameters for scripts in README

Commit:
Merge branch 'rStudioDemoCourse' of github.com:SDU-eScience/UCloud-Courses into rStudioDemoCourse

Commit:
move README to course's top level folder

Rymik19 committed:
adds instructions for testing image and configuring yml files

Rymik19 committed:
app yml file with start script options

Rymik19 committed:
integrates docker run with build script

Rymik19 committed:
helper command for using build-and-run script

Commit:
Revert "integrates docker run with build script"

This reverts commit d97f49f40e29b3f37bc99034b0db6538027e2d45.

Commit:
change hyphens in names of python scripts to enable importing these scripts

Commit:
script for building and running in process

Commit:
script for building and running in process

Commit:
resolve pseudoconfilct

Commit:
rename r-demo-course to r-studio-demo-course

Commit:
meeting

Rymik19 committed:
updates build and run script for rstudio

JHavstein committed:
Incorporate comments from code reveiw

Rymik19 committed:
changed location and name for run container script

Rymik19 committed:
reverts changing hyphens for importing scripts

Rymik19 committed:
Add build-and-run-course_template and integrate with create-new-course.py

Rymik19 committed:
updates documentation

Rymik19 committed:
updates README.md

Commit:
introduce redownload flag and rename yml-files to include coursecode 0000

Rymik19 committed:
fixes documentation

Rymik19 committed:
add start command to app yaml file

Rymik19 committed:
removes build-and-run-course_template; adds run-docker-container.py

Rymik19 committed:
adds startCommand required for run-docker-container.py

Rymik19 committed:
script comment fix

Rymik19 committed:
various comment and variable changes

Commit:
Minor corrections to the README

Commit:
Add description to Dockerfile and yml-files, set values in yml files.

JHavstein committed:
Using Docker from course folders

Started implemented a more user-friendly solution. Templates are copied into the course folders. These have bare minimum requirements for the teacher to fill out. These call the main scripts for building images and running containers located in `scripts/`.

JHavstein committed:
Temporary modifications

Will need some final modifications in the end.

JHavstein committed:
Continue working on scripts/templates

JHavstein committed:
Continue

JHavstein committed:
Finalize scripts

JHavstein committed:
Extra spacing

JHavstein committed:
More string parsing

JHavstein committed:
Minor things

JHavstein committed:
Remove possibility of `build-args`

JHavstein committed:
Remove excess spacing

JHavstein committed:
Fix comment alignment

JHavstein committed:
Rename files

From *`.py` to `*.template`

JHavstein committed:
Update yaml templates

JHavstein committed:
Change scripts for building/running

JHavstein committed:
Update create-new-course.py

JHavstein committed:
Update docker-build.template

Rymik19 committed:
changes force_download flag to redownload flag

Rymik19 committed:
redownload flag fix

JHavstein committed:
Add maintainer info

JHavstein committed:
Minor changes to templates

JHavstein committed:
Rename 'Courses/Other' -> 'Courses/OTHER'

Makes parsing of paths between course folders and scripts folder easier.

JHavstein committed:
Add possibility of start command

The teacher can now specify a start command for a Docker container in `docker-run.py` (default: `start_course`).

Rymik19 committed:
removes unnecessary files and startCommand from yml

Rymik19 committed:
Merge branch 'add-course-setup' into rStudioDemoCourse

Rymik19 committed:
adapts changes from mergin add-course-setup branch

Rymik19 committed:
adds start_command

Rymik19 committed:
doc update

Rymik19 committed:
Fixes start_course.sh script

Rymik19 committed:
fix Initialization section links

Rymik19 committed:
remove root access in dockerfile

Rymik19 committed:
merge main into rStudioDemoCourse

Rymik19 committed:
add minor changes from templates to course files.

Rymik19 committed:
minor fixes to .yml files

Rymik19 committed:
minor doc changes

Rymik19 committed:
misc changes

JHavstein committed:
Merge branch 'main' into rStudioDemoCourse

JHavstein reviewed (changes requested):
I haven't reviewed the files i `renv/` and `slides/` since I assume that these originate from somewhere else. Let me know if I should have a look at these files also. 

@JHavstein They come from the course repository: https://github.com/jeselginAU/demo-r-studio-course/tree/main/renv 
I have tested that the Rmd files can be "Knitted" and that the `renv` environment is loaded using the `activate.R` to manage R package dependencies. -Rytis

_General changes:_ 

- The course code should be changed everywhere (in file names, folder names, and inside files) from `0000` to the AU actual course code.

@JHavstein  I believe this course is only supposed to be an example on how to set up a course, so there is no actual AU course code. -Rytis

@Rymik19 @jeselginAU OK, let's keep `0000` as the course code then. -Jonas 

- Likewise, the course release date should be changed from `2024-01-01` to the correct one, say `2024-01-09`.

_Testing:_ 

I've only tested locally so far. The image builds successfully with `docker-build.py` and the different classes can be launched correctly using `docker-run.py` and editing `START_COMMAND`. (The container won't run interactively, but I think this has to do with the base image.)

Rymik19 committed:
PR changes

Rymik19 committed:
Doc PR change

Commit:
update README

Commit:
re-enable apt-get safeguards

JHavstein committed:
Minor updates

`Conda` is not available in the container.

`-p` is not implemented in `start_course.sh`.

JHavstein committed:
Set `runAsRoot: true`

JHavstein reviewed (changes requested):
If there are any files or folders that aren't used (`activate.R`?) they should be removed.

The course works on dev.

Rymik19 committed:
remove req.sh

Rymik19 committed:
Update Courses/AU/r-studio-demo-course__0000/README.md

Co-authored-by: Jonas <33627730+JHavstein@users.noreply.github.com>

Rymik19 committed:
documentation fixes.

Rymik19 committed:
nlp course documentation update

Commit:
Revert "nlp course documentation update" (belongs on its own branch)

This reverts commit b60e38e38c754d8dbe53587029223a702b36030c.

Rymik19 committed:
adding new course repository and environment activation

Rymik19 commented:
I have created a simple script `analysis.R` that uses `dataset.csv` to plot some data (Math scores) using `ggplot2` library for plotting and `dplyr` library for summarizing data. I have made it modular to feature functions and a main function. It prints out some statistics about the data and creates plots for distribution and density.

It is available at the course material repository [here](https://github.com/Rymik19/rStudioDemoCourse). 

The libraries are installed using an R environment. The environment files (`renv` folder, `.Rprofile` and `renv.lock`) are added in the Docker image at `/home/${USERID}/Renvironment`. At startup, they are then copied to `/work` and the environment is activated using `Rscript /work/Renvironment/renv/activate.R`.

@JHavstein and @emolinaro I would appreciate your feedback on this approach.

JHavstein commented:
> @JHavstein and @emolinaro I would appreciate your feedback on this approach.

@Rymik19 I'll review it as soon as possible.

JHavstein commented on Courses/AU/r-studio-demo-course__0000/2024-09-01/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+## Add environment files
+COPY --chown=${USERID}:${GROUPID} renv /Renvironment/renv
+COPY --chown=${USERID}:${GROUPID} .Rprofile /Renvironment/.Rprofile
+COPY --chown=${USERID}:${GROUPID} renv.lock /Renvironment/renv.lock
```
Remove the `renv`setup. 
```suggestion
```

JHavstein commented on Courses/AU/r-studio-demo-course__0000/2024-09-01/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+## Use bash instead of sh
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+
+WORKDIR /home/${USERID}
```
```suggestion
WORKDIR /tmp
```

Rymik19 replied:
I think this command is not necessary.

JHavstein commented on Courses/AU/r-studio-demo-course__0000/2024-09-01/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+
+WORKDIR /home/${USERID}
+
```
Install the required R packages directly from the Dockerfile like so. This removes the need for an R environment:
```suggestion
RUN sudo apt-get update -qq \
 && install2.r --error --deps TRUE \
    ggplot2 \
    dplyr \
 && rm -rf /tmp/downloaded_packages/ /tmp/*.rds \
 && sudo apt-get clean \
 && sudo rm -rf /var/lib/apt/lists/*
```
In fact, this step is not necessary since the entire `tidyverse` package suite is already installed into the base image. But I think it still could make sense to include the code above for teachers to see how to install R packages from a Dockerfile - although perhaps we should choose to install some package which is not pre-installed in the base image.

Rymik19 replied:
I will test the class script without installing the packages to confirm `tidyverse` is enough. I will then change the script to use other packages not included in the base image.

JHavstein replied:
OK. 

If you prepare a script for `class_02`, can you use a package in that script that is not present in the base image and then install the package from the Dockerfile? Just to demonstrate how this is done for people who will be using this course for inspiration.

JHavstein commented on Courses/AU/r-studio-demo-course__0000/2024-09-01/.Rprofile:
Can be deleted.

JHavstein commented on Courses/AU/r-studio-demo-course__0000/2024-09-01/renv.lock:
Can be deleted.

JHavstein commented on Courses/AU/r-studio-demo-course__0000/2024-09-01/renv/activate.R:
Can be deleted.

JHavstein commented on Courses/AU/r-studio-demo-course__0000/2024-09-01/renv/settings.json:
Can be deleted.

JHavstein commented on Courses/AU/r-studio-demo-course__0000/2024-09-01/start_course.sh:
```diff
@@ -0,0 +1,92 @@
+  exit 1
+fi
+
+echo "Environment successfully activated!"
```
```suggestion
```

JHavstein commented on Courses/AU/r-studio-demo-course__0000/README.md:
This README must be revised such that it only contains information which is relevant specifically for this course. As such, it can probably be kept pretty short. For general things about how to develop UCloud courses etc., teachers should read the repo's wiki.

@JHavstein I have updated the documentation. It borrows a few paragraphs from the NLP course README regarding machine and class selection, which could just be referenced. It highlights the scripts used in `class_01` and `class_02`, with the latter requiring package installation in the Dockerfile. /Rytis

JHavstein commented on Courses/AU/r-studio-demo-course__0000/2024-09-01/renv/.gitignore:
Can be deleted.

JHavstein commented on Courses/AU/r-studio-demo-course__0000/2024-09-01/r-studio-demo-course__0000-app.yml:
```diff
@@ -0,0 +1,83 @@
+        name: class 1
+        value: "class_01"
+      -
+        name: class 2
+        value: "class_02"
+      -
+        name: class 3
+        value: "class_03"
+      -
+        name: class 4
+        value: "class_04"
```
As far as I can see, only `class_01` is available in the external repo. Is this intentional? If so, we should remove the other classes from here.

Rymik19 replied:
I have only created a script for one class, should I create another one for class_02 to facilitate class selection? I will remove the missing classes.

JHavstein replied:
Yes, please create one for `class_02` as well so that it will be possible to select between these class 1 and class 2. It can just be very simple script. When preparing the script, see also this comment: https://github.com/SDU-eScience/UCloud-Courses/pull/9#discussion_r1910366959 from above.

You can remove class 3 and 4 from the yaml file.

JHavstein commented on Courses/AU/r-studio-demo-course__0000/2024-09-01/start_course.sh:
```diff
@@ -0,0 +1,92 @@
+    printf "\n== Start RStudio ==\n"
+    printf "===================\n\n"
+    sudo /init  
+fi
```
```suggestion
fi

```

JHavstein reviewed (changes requested):
The course container works fine on dev. But the environment does not work seamlessly. 

I understand the general appeal of working inside a dedicated R environment, but I believe that the (simpler) solution of installing the required R packages from the Dockerfile (see review comments) makes more sense in this case. By installing R packages in this way we also avoid cluttering the `/work` folder, and the different course development scripts become easier to read which makes sense given that this is meant to be a demo course.

Rymik19 committed:
removes environment files

Rymik19 commented on Courses/AU/r-studio-demo-course__0000/2024-09-01/docker-run.py:
@JHavstein I experienced some issues when testing the -it flag.

When setting INTERACTIVE_MODE = "True":

[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...s6-chown: fatal: unable to chown /var/run/s6/etc/cont-init.d/02_userconf: Operation not permitted
s6-chown: fatal: unable to chown /var/run/s6/etc/cont-init.d/01_set_env: Operation not permitted
s6-chmod: fatal: unable to change mode of /var/run/s6/etc/cont-init.d/01_set_env: Operation not permitted
s6-chmod: fatal: unable to change mode of /var/run/s6/etc/cont-init.d/02_userconf: Operation not permitted
s6-chown: fatal: unable to chown /var/run/s6/etc/services.d/rstudio/run: Operation not permitted
s6-chmod: fatal: unable to change mode of /var/run/s6/etc/services.d/rstudio/run: Operation not permitted
s6-chown: fatal: unable to chown /var/run/s6/etc/services.d/rstudio/finish: Operation not permitted
s6-chmod: fatal: unable to change mode of /var/run/s6/etc/services.d/rstudio/finish: Operation not permitted
exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 01_set_env: executing... 
foreground: warning: unable to spawn /var/run/s6/etc/cont-init.d/01_set_env: Permission denied
[cont-init.d] 01_set_env: exited 127.
[cont-init.d] 02_userconf: executing... 
foreground: warning: unable to spawn /var/run/s6/etc/cont-init.d/02_userconf: Permission denied
[cont-init.d] 02_userconf: exited 127.
[cont-init.d] done.
[services.d] starting services
s6-supervise (child): fatal: unable to exec run: Permission denied
s6-supervise rstudio: warning: unable to spawn ./run - waiting 10 seconds
[services.d] done.
s6-supervise (child): fatal: unable to exec run: Permission denied
s6-supervise rstudio: warning: unable to spawn ./run - waiting 10 seconds
^C[cont-finish.d] executing container finish scripts...
[cont-finish.d] done.


Any ideas why this happens?

JHavstein replied:
I'll have to look more into it. But as far as I know, this is caused by the base image which is `rocker/rstudio` and not one of our own linux base images. As far as I can tell it has nothing to do with the scripts in this repo.

Rymik19 committed:
removes environment initialization

Rymik19 commented:
~~@JHavstein One other issue I am experiencing now is that the downloaded repository files are in the container at `/work/classes_01` however they do not appear in the RStudio interface. I do not know why, but perhaps there is an issue with my local setup. 
Could you verify that by testing if the downloaded repository files (`analysis.R` & `dataset.csv`) appear in R Studio?~~

@Rymik19 I'll do that first thing when I put a new version on dev for testing. Please let me know when the code is ready for a new iteration of testing. (/Jonas)

@JHavstein I have created a new `class_02` folder with a [script](https://github.com/Rymik19/rStudioDemoCourse/blob/main/classes/class_02/example_with_new_packages.R) that uses `janitor` and `plotly` packages. I have checked that these are not included in the base image. I have tested this locally with success. My issue above was tied with using WSL engine directly, instead of Docker Desktop when testing, which I solved.

Rymik19 committed:
install packages for class_02 R script

Rymik19 committed:
add downloaded course material path to print

Rymik19 committed:
adds empty line

Rymik19 committed:
remove irrelevant class options

Rymik19 committed:
update documentation

JHavstein commented on Courses/AU/r-studio-demo-course__0000/README.md:
```diff
@@ -0,0 +1,49 @@
+# RStudio Demo Course
+
+This course serves as an example of setting up a course with `RStudio`.
```
```suggestion
This course is a demonstration of how to set up a simple UCloud course with `RStudio`.
```

JHavstein commented on Courses/AU/r-studio-demo-course__0000/README.md:
```diff
@@ -0,0 +1,49 @@
+
+### Select Course Module
+
+Choose the relevant class (module) from the dropdown menu in the UCloud job configuration interface.
```
```suggestion
Choose the relevant class (module) from the dropdown menu in the UCloud job submission page.
```

JHavstein commented on Courses/AU/r-studio-demo-course__0000/README.md:
```diff
@@ -0,0 +1,49 @@
+### Select Course Module
+
+Choose the relevant class (module) from the dropdown menu in the UCloud job configuration interface.
+The course materials (instructions and datasets) are hosted in a [GitHub](https://github.com/Rymik19/rStudioDemoCourse/tree/main/classes) repository.
```
```suggestion
The course materials (instructions and datasets) are hosted in an external [GitHub](https://github.com/Rymik19/rStudioDemoCourse/tree/main/classes) repository.
```

JHavstein commented on Courses/AU/r-studio-demo-course__0000/README.md:
```diff
@@ -0,0 +1,49 @@
+Choose the relevant class (module) from the dropdown menu in the UCloud job configuration interface.
+The course materials (instructions and datasets) are hosted in a [GitHub](https://github.com/Rymik19/rStudioDemoCourse/tree/main/classes) repository.
+
+The `analysis.R` script for `class_01` demonstrates data validation, group-wise summarization, and visualization using `ggplot2` which is already included in the base `RStudio` image.
```
```suggestion
The `analysis.R` script for `class_01` demonstrates data validation, group-wise summarization, and visualization using the `ggplot2` library.
```

JHavstein commented on Courses/AU/r-studio-demo-course__0000/README.md:
```diff
@@ -0,0 +1,49 @@
+
+Once you start the course app, the course materials for the selected course module will be downloaded to the path `/work/class_[module#]`.
+
+When the job is completed, the class_[module#] folder will be available in your UCloud drive inside the [Jobs](https://docs.cloud.sdu.dk/guide/monitoring.html#job-completed) folder.
```
```suggestion
When the job is completed, the `class_<module#>` folder will be available in your UCloud drive inside the [Jobs](https://docs.cloud.sdu.dk/guide/monitoring.html#job-completed) folder.
```

JHavstein commented on Courses/AU/r-studio-demo-course__0000/README.md:
```diff
@@ -0,0 +1,49 @@
+
+### Re-download Course Files
+
+By default, this parameter is set to `false` meaning the course materials are downloaded, except if a folder called `/work/class_[module#]` already exists.
```
```suggestion
By default, this optional parameter is set to `false` meaning the course materials are downloaded fro mthe external GitHub repository, except if a folder called `/work/class_[module#]` already exists.
```

JHavstein commented on Courses/AU/r-studio-demo-course__0000/README.md:
```diff
@@ -0,0 +1,49 @@
+In other words, if you select class 1 and you mount a folder called `class_01` by using the parameter _Select folders to use_,
+files from the teachers GitHub repos won't be re-downloaded.
+
+You can overwrite the existing folder by setting the option _Re-download course files_  to `true`. This will overwrite `/work/class_[module#]` inside the job and the mounted UCloud storage.
```
```suggestion
You can overwrite the existing folder by setting the option _Re-download course files_  to `true`. This will overwrite `/work/class_[module#]` with the current contents of the folder in the external GitHub repository.
```

JHavstein replied:
```suggestion
You can overwrite the existing folder by setting the option _Re-download course files_  to `true`. This will overwrite `/work/class_<module#>` with the newest version of the relevant course materials from the external GitHub repository.
```

JHavstein commented on Courses/AU/r-studio-demo-course__0000/README.md:
```diff
@@ -0,0 +1,49 @@
+
+### Initialization
+
+For information on how to use the _Initialization_ parameter, please refer to the [Initialization - Bash script](https://docs.cloud.sdu.dk/hands-on/init-sh.html), [Initialization - Conda packages](https://docs.cloud.sdu.dk/hands-on/init-conda.html), and [Initialization - pip packages](https://docs.cloud.sdu.dk/hands-on/init-pip.html) section of the documentation.
```
This can be deleted since it's a general feature when submitting UCloud jobs.

```suggestion

```

JHavstein commented on Courses/AU/r-studio-demo-course__0000/2024-09-01/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+
+## Use bash instead of sh
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+
```
```suggestion

## Install required R libraries
```

JHavstein commented on Courses/AU/r-studio-demo-course__0000/2024-09-01/r-studio-demo-course__0000-app.yml:
```diff
@@ -0,0 +1,77 @@
+    type: boolean
+    description: >
+      Select to download course module files. By default, this option is set to false, meaning that if a folder for the selected class already exists, no new files from the external repo are downloaded. 
+      If set to true, any existing files in your class folder will be overwritten.
```
To keep the description short.

```suggestion
      Re-download course materials from external repository (default: false). If set to true, the contents of mounted class folders for the selected class will be overwritten.
```

JHavstein commented on Courses/AU/r-studio-demo-course__0000/README.md:
```diff
@@ -0,0 +1,49 @@
+From the job submission [page](https://cloud.sdu.dk/app/jobs/create?app=nlp-demo-course__147222U005),
+choose machines with only vCPU resources, such as `u1-standard-4` and `uc1-gc1-4` as they provide the minimum required resources for the course materials.
+
+You can select the machine type from the dropdown menu when configuring your UCloud job.
```
This can be deleted since it's a general thing when submitting jobs on UCloud.
```suggestion
```

JHavstein commented on Courses/AU/r-studio-demo-course__0000/README.md:
```diff
@@ -0,0 +1,49 @@
+
+The `example_with_new_packages.R` script for `class_02` uses `janitor` for data cleaning and `plotly` for creating interactive visualizations. These packages need to be installed in the `Dockerfile` to ensure they are pre-configured and readily available in RStudio without requiring manual installation by the user.
+
+Once you start the course app, the course materials for the selected course module will be downloaded to the path `/work/class_[module#]`.
```
```suggestion
Once you start the course app, the course materials for the selected course module will be downloaded to the path `/work/class_<module#>`.
```

JHavstein commented on Courses/AU/r-studio-demo-course__0000/README.md:
```diff
@@ -0,0 +1,49 @@
+
+By default, this parameter is set to `false` meaning the course materials are downloaded, except if a folder called `/work/class_[module#]` already exists.
+In other words, if you select class 1 and you mount a folder called `class_01` by using the parameter _Select folders to use_,
+files from the teachers GitHub repos won't be re-downloaded.
```
```suggestion
In other words, if you select class 1 and you mount a folder called `class_01` by using the parameter _Select folders to use_, files from the external GitHub repository won't be re-downloaded.
```

JHavstein commented on Courses/AU/r-studio-demo-course__0000/2024-09-01/start_course.sh:
```diff
@@ -0,0 +1,75 @@
+    printf "\n== Start RStudio ==\n"
+    printf "===================\n\n"
+    sudo /init  
+fi
```
This logic needs to be changed a bit. Specifically, it doesn't handle the case where a class module is selected, a class folder is mounted, and redownload is set to `false`. 

As I see it, the following cases must be handled:

1.  No class selected -> Start RStudio immediately
2. Class module selected AND no class folder mounted -> Download course materials from repo and start RStudio 
3. Class module selected AND a class folder mounted -> Download course materials according to redownload flag and start RStudio

I think the best way to do it would be to re-write the logic to first handle the above three cases and then move 
```
printf "\n==================="
printf "\n== Start RStudio ==\n"
printf "===================\n\n"
sudo /init  
```
out of the conditional statements to ensure that it's always started.

Rymik19 replied:
I have implemented this change. It is challenging for me to test locally the cases where a class folder is mounted, could you please ensure it is handled correctly?

JHavstein replied:
Great - I'll be sure to test it.

I also just realized that `Class` is a mandatory parameter which defaults to `class_01`. This means that case 1 from my previous comment is not relevant after all. I'll just change this in the script (it is essentially just 2-3 lines that need to be deleted).

JHavstein reviewed (changes requested):
The course works on UCloud - but see the comment to `start_course.sh`.

Rymik19 committed:
Update Courses/AU/r-studio-demo-course__0000/README.md

Co-authored-by: Jonas <33627730+JHavstein@users.noreply.github.com>

Rymik19 committed:
Update Courses/AU/r-studio-demo-course__0000/README.md

Co-authored-by: Jonas <33627730+JHavstein@users.noreply.github.com>

Rymik19 committed:
Update Courses/AU/r-studio-demo-course__0000/README.md

Co-authored-by: Jonas <33627730+JHavstein@users.noreply.github.com>

Rymik19 committed:
Update Courses/AU/r-studio-demo-course__0000/README.md

Co-authored-by: Jonas <33627730+JHavstein@users.noreply.github.com>

Rymik19 committed:
Update Courses/AU/r-studio-demo-course__0000/README.md

Co-authored-by: Jonas <33627730+JHavstein@users.noreply.github.com>

Rymik19 committed:
Update Courses/AU/r-studio-demo-course__0000/README.md

Co-authored-by: Jonas <33627730+JHavstein@users.noreply.github.com>

Rymik19 committed:
Update Courses/AU/r-studio-demo-course__0000/README.md

Co-authored-by: Jonas <33627730+JHavstein@users.noreply.github.com>

Rymik19 committed:
Update Courses/AU/r-studio-demo-course__0000/README.md

Co-authored-by: Jonas <33627730+JHavstein@users.noreply.github.com>

Rymik19 committed:
Update Courses/AU/r-studio-demo-course__0000/README.md

Co-authored-by: Jonas <33627730+JHavstein@users.noreply.github.com>

Rymik19 committed:
handles missing case in class module logic

JHavstein committed:
Update Dockerfile

JHavstein committed:
Update Dockerfile

JHavstein committed:
Minor changes

JHavstein committed:
Typo

JHavstein commented:
@Rymik19: I observed the following: 

I selected a class, and mounted a class folder from a previous job where I had added some files that were not a part of the downloaded course material. Finally, I set the `re-download` flag to `true`. This resulted in the course files being re-downloaded, as expected. But the files I had added to the mounted course folder were still present after the re-download (i.e., along with the re-downloaded course material files). 

Is this the intended behavior? Or should the mounted course folder be wiped of content before the course files are redownload into the folder? 

_Follow-up:_ @Rymik19 If it's the intended behavior the section about the redownload flag in the `README` should be adjusted accordingly. If it's not the intended behavior, `start_course.sh` should be adjusted accordingly. Otherwise, I think the PR is ready to be approved (still need to do final testing).

@Rymik19 Let's stick with the current behavior. I'll make the changes in the `README` and approve the PR today or tomorrow.

@JHavstein I am sorry for the late reply. It is intended behavior, it can help with backing up course files by renaming them. Files can also be removed through UCloud UI before mounting the folder, so there should be no issue.

JHavstein committed:
Update README

JHavstein committed:
Update description of re-download flag

JHavstein committed:
Minor

JHavstein committed:
Merge branch 'main' into rStudioDemoCourse

JHavstein committed:
Typo

JHavstein committed:
Typo
